File: packages/sandboxd/src/egress_proxy/logging.rs
use serde_json::{Map, Value};

use crate::supervision::SupervisedComponent;
use crate::time::{Clock, format_rfc3339_timestamp};

#[derive(Clone, Copy)]
pub(super) struct EgressProxyLogContext<'a> {
    pub(super) clock: &'a dyn Clock,
    pub(super) sandbox_instance_id: &'a str,
}

pub(super) fn emit_egress_proxy_log(
    clock: &dyn Clock,
    sandbox_instance_id: &str,
    event: &str,
    extra_fields: &[(&str, Value)],
) {
    if let Some(line) =
        serialize_egress_proxy_log_line(clock, sandbox_instance_id, event, extra_fields)
    {
        eprintln!("{line}");
    }
}

pub(super) fn serialize_egress_proxy_log_line(
    clock: &dyn Clock,
    sandbox_instance_id: &str,
    event: &str,
    extra_fields: &[(&str, Value)],
) -> Option<String> {
    let observed_at = format_rfc3339_timestamp(clock.now_system_time()).ok()?;
    let mut payload = Map::new();
    payload.insert("event".to_string(), Value::String(event.to_string()));
    payload.insert(
        "sandboxInstanceId".to_string(),
        Value::String(sandbox_instance_id.to_string()),
    );
    payload.insert(
        "component".to_string(),
        Value::String(SupervisedComponent::EgressProxy.as_str().to_string()),
    );
    payload.insert("observedAt".to_string(), Value::String(observed_at));
    for (field_name, field_value) in extra_fields {
        payload.insert((*field_name).to_string(), field_value.clone());
    }
    serde_json::to_string(&Value::Object(payload)).ok()
}

File: packages/sandboxd/egressproxy/logging.go
package egressproxy

import (
	"encoding/json"
	"fmt"
	"io"
	"os"

	"github.com/mistle/sandboxd/supervision"
	"github.com/mistle/sandboxd/timeutil"
)

type LogContext struct {
	Clock             timeutil.Clock
	SandboxInstanceID string
}

func EmitLog(clock timeutil.Clock, sandboxInstanceID string, event string, extraFields map[string]any) error {
	return EmitLogTo(os.Stderr, clock, sandboxInstanceID, event, extraFields)
}

func EmitLogTo(writer io.Writer, clock timeutil.Clock, sandboxInstanceID string, event string, extraFields map[string]any) error {
	line, err := SerializeLogLine(clock, sandboxInstanceID, event, extraFields)
	if err != nil {
		return err
	}
	if _, err := fmt.Fprintln(writer, line); err != nil {
		return fmt.Errorf("failed to write egress proxy log line: %w", err)
	}
	return nil
}

func SerializeLogLine(clock timeutil.Clock, sandboxInstanceID string, event string, extraFields map[string]any) (string, error) {
	if clock == nil {
		return "", fmt.Errorf("egress proxy log clock is required")
	}

	payload := map[string]any{
		"event":             event,
		"sandboxInstanceId": sandboxInstanceID,
		"component":         string(supervision.ComponentEgressProxy),
		"observedAt":        timeutil.FormatRFC3339Timestamp(clock.NowSystemTime()),
	}
	for fieldName, fieldValue := range extraFields {
		payload[fieldName] = fieldValue
	}

	serialized, err := json.Marshal(payload)
	if err != nil {
		return "", fmt.Errorf("failed to serialize egress proxy log line: %w", err)
	}
	return string(serialized), nil
}
